drivers/sensor/scd30/scd30.c

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2021 Advanced Climate Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define DT_DRV_COMPAT sensirion_scd30
+
+#include <device.h>
+#include <drivers/i2c.h>
+#include <kernel.h>
+#include <drivers/sensor.h>
+#include <sys/__assert.h>
+#include <sys/byteorder.h>
+#include <sys/crc.h>
+#include <logging/log.h>
+LOG_MODULE_REGISTER(SCD30, CONFIG_SCD30_LOG_LEVEL);
+
+#include "scd30.h"
+
+/**
+ * @brief Take the received  word (16 bit), if the data is valid, add it to dst
+ *
+ * @param data data received from sensor
+ * @param dst buffer to store data
+ * @return int
+ */
+static int scd30_fill_data_buf(uint8_t *data, uint8_t *dst)
+{
+	if (scd30_check_crc(data, 2, data[2])) {
+		return -EIO;
+	}
+
+	dst[0] = data[0];
+	dst[1] = data[1];
+
+	return 0;
+}
+
+float scd30_bytes_to_float(const uint8_t *bytes)
+{
+	union {
+		uint32_t u32_value;
+		float float32;
+	} tmp;
+
+	tmp.u32_value = sys_get_be32(bytes);
+	return tmp.float32;
+}
+
+static int scd30_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	uint8_t rx_buf[6 * (SCD30_WORD_SIZE + SCD30_CRC8_LEN)] = {};
+	uint8_t data_buf[6 * SCD30_WORD_SIZE] = {};
+	struct scd30_data *data = dev->data;
+	const struct scd30_config *cfg = dev->config;
+	int rc = 0;
+
+	rc = scd30_write_command(dev, SCD30_CMD_GET_DATA_READY);
+	if (rc != 0) {
```

Comment:
this function should check if requested channel is supported (or `SENSOR_CHAN_ALL`)

Reply:
One question here. If the sensor itself only supports reading all data at once, what would be the correct way of handling this? 
1) regardless of which channel is asked for, updating all channels or,
2) drop data which is not asked for and return `ENODATA` when the user asks to fetch a different channel within the sample period of the sensor.
After my latest the driver does the first option. I couldn't find the preferred way of handling this in the docs.

Reply:
you may take this as an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/ina23x/ina23x.c#L156

Reply:
The difference between the two sensors here, as far as I can tell, is that the ina23x supports reads for the separate channels it measures, so performing a call like `sensor_sample_fetch_chan(dev, SENSOR_CHAN_VOLTAGE)` leaves the data
for the other two channels intact in the sensor so they can be read a later stage.
So code that looks like the below snippet, will work as expected, i.e. it will update the voltage and current returned by
`sensor_channel_get()`, while leaving power untouched.
```c
sensor_sample_fetch_chan(dev, SENSOR_CHAN_VOLTAGE);
sensor_sample_fetch_chan(dev, SENSOR_CHAN_CURRENT);
```

For the SCD30 this is not the case, the sensor only supports a readout of all measurement registers at once.
This leads to the situation where perform the following two calls leads to behavior that is unexpected.
```c
sensor_sample_fetch_chan(dev, SENSOR_CHAN_CO2);
sensor_sample_fetch_chan(dev, SENSOR_CHAN_AMBIENT_TEMP);
```
The way I read the [datasheet](https://cdn.sparkfun.com/assets/d/c/0/7/2/SCD30_Interface_Description.pdf) (sections 1.3.3 and 1.3.4) this would cause the second call to `sensor_sample_fetch_chan()` to fail because no new data is available.

Reply:
Thanks for the detailed explanation, in such case, you should likely validate that `chan == SENSOR_CHAN_ALL`.